services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractRegistrationService.java

```diff
@@ -159,6 +276,38 @@ public final void setTenantInformationService(final TenantInformationService ten
         return assertRegistration(tenantId, deviceId, gatewayId, NoopSpan.INSTANCE);
     }
 
+    /**
+     * Add a new edge device in the process of auto-provisioning of not yet registered edge devices sending data
+     * via a gateway.
+     *
+     * @param tenantId The tenant that the device belongs to.
+     * @param deviceId The ID the device should be registered under.
+     * @param device Device information, must not be {@code null}.
+     * @param span The active OpenTracing span for this operation.
+     *
+     * @return A future indicating the outcome of the operation.
+     */
+    protected abstract Future<OperationResult<Id>> addEdgeDevice(String tenantId, Optional<String> deviceId, Device device, Span span);
+
+    private Future<OperationResult<Id>> setAutoProvisioningNotificationSent(final String tenantId, final String deviceId, final JsonObject deviceJson, final Span span) {
+        final Device device = deviceJson.mapTo(Device.class);
+        device.setAutoProvisioningNotificationSent(true);
+        return updateDevice(tenantId, deviceId, device, Optional.empty(), span);
+    }
+
+    /**
+     * Updates device registration data.
+     *
+     * @param tenantId The tenant the device belongs to.
+     * @param deviceId The ID of the device to update the registration for.
+     * @param device Device information, must not be {@code null}.
+     * @param resourceVersion The identifier of the resource version to update.
+     * @param span The active OpenTracing span for this operation.
+     *
+     * @return A future indicating the outcome of the operation.
```

Comment:
What will be contained in the future if the operation fails? Why should it fail?

Reply:
Fixed.